indico/modules/search/controllers.py

```diff
@@ -84,50 +69,79 @@ def _process(self):
 
 
 class InternalSearch(IndicoSearchProvider):
-    def search(self, query, access, page=1, object_types=(), **params):
+    def search(self, query, user, page=None, object_types=(), **params):
         if object_types == [SearchTarget.category]:
-            total, results = InternalSearch.search_categories(page, query, params.get('category_id'))
+            pagenav, results = self.search_categories(query, user, page, params.get('category_id'))
         elif object_types == [SearchTarget.event]:
-            total, results = InternalSearch.search_events(page, query, params.get('category_id'))
+            pagenav, results = self.search_events(query, user, page, params.get('category_id'))
         else:
-            total, results = 0, []
+            pagenav, results = {}, []
         return {
-            'total': total,
-            'pages': math.ceil(total / self.RESULTS_PER_PAGE),
+            'total': -1 if results else 0,
+            'pagenav': pagenav,
             'results': results,
         }
 
-    @staticmethod
-    def search_categories(page, q, category_id):
+    def _paginate(self, query, page, column, user):
+        reverse = False
+        pagenav = {'prev': None, 'next': None}
+        if not page:
+            query = query.order_by(column.desc())
+        elif page > 0:  # next page
+            query = query.filter(column < page).order_by(column.desc())
+            # since we asked for a next page we know that a previous page exists
+            pagenav['prev'] = -(page - 1)
+        elif page < 0:  # prev page
+            query = query.filter(column > -page).order_by(column)
+            # since we asked for a previous page, wek now that a next page exists
+            pagenav['next'] = -(page - 1)
+            reverse = True
+
+        def _can_access(obj):
+            return obj.effective_protection_mode == ProtectionMode.public or obj.can_access(user, allow_admin=False)
+
+        res = get_n_matching(query, self.RESULTS_PER_PAGE + 1, _can_access, prefetch_factor=20)
+
+        if len(res) > self.RESULTS_PER_PAGE:
+            # we queried 1 more so we can see if there are more results available
+            assert len(res) == self.RESULTS_PER_PAGE + 1
+            del res[-1]
```

Comment:
Feel free to ignore this if you prefer it this way. But we could in principle just trim it as it seems harmless.

```suggestion
            del res[self.RESULTS_PER_PAGE:]
```

Reply:
yeah, i thought the same, but i was too lazy to test the alternative ;) (i always want to be sure to avoid off-by-one errors when using slices)